### PR TITLE
Add more blend equations from 3dbrew

### PR DESCRIPTION
**Pull request by Alegend45**

### CODE REVIEW

File: src/video_core/pica.h [hunk 1]
```diff
@@ -339,6 +339,10 @@ struct Regs {
         union {
             enum BlendEquation : u32 {
                 Add = 0,
+                Subtract = 1,
```
Comment (neobrain) [inline src/video_core/pica.h]: Please align these at `=`.

File: src/video_core/rasterizer.cpp [hunk 2]
```diff
@@ -627,6 +627,46 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
                     combiner_output = result.Cast<u8>();
                     break;
                 }
+                
+                case params.Subtract:
+                {
+                    auto result = (combiner_output * srcfactor - dest * dstfactor) / 255;
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: Could you precompute `combiner_output * srcfactor` and `dest * dstfactor` outside of the switch to reduce redundancy?

File: src/video_core/rasterizer.cpp [hunk 3]
```diff
@@ -627,6 +627,46 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
...
+                    result.r() = std::min(255, result.r());
+                    result.g() = std::min(255, result.g());
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: This should be clamped against zero instead.

File: src/video_core/rasterizer.cpp [hunk 4]
```diff
@@ -627,6 +627,46 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
...
+                    result.b() = std::min(255, result.b());
+                    combiner_output = result.Cast<u8>();
+                    break;
+                }
+                
+                case params.ReverseSubtract:
+                {
+                    auto result = (dest * dstfactor - combiner_output * srcfactor) / 255;
+                    result.r() = std::min(255, result.r());
+                    result.g() = std::min(255, result.g());
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: this should be clamped against zero instead.

File: src/video_core/rasterizer.cpp [hunk 5]
```diff
@@ -627,6 +627,46 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
...
+                    result.b() = std::min(255, result.b());
+                    combiner_output = result.Cast<u8>();
+                    break;
+                }
+                
+                case params.Min:
+                {
+                    auto result = std::min(combiner_output * srcfactor, dest * dstfactor) / 255;
+                    result.r() = std::min(255, result.r());
+                    result.g() = std::min(255, result.g());
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: This need not be clamped

File: src/video_core/rasterizer.cpp [hunk 6]
```diff
@@ -627,6 +627,46 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
...
+                    combiner_output = result.Cast<u8>();
+                    break;
+                }
+                
+                case params.Max:
+                {
+                    auto result = std::max(combiner_output * srcfactor, dest * dstfactor) / 255;
+                    result.r() = std::min(255, result.r());
+                    result.g() = std::min(255, result.g());
+                    result.b() = std::min(255, result.b());
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: This need not be clamped

**Comment by neobrain** [comment]
Made some comments, other than those, LGTM.

File: src/video_core/rasterizer.cpp [hunk 7]
```diff
@@ -616,17 +616,54 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
                                                LookupFactorA(params.factor_source_a));
                 auto dstfactor = Math::MakeVec(LookupFactorRGB(params.factor_dest_rgb),
                                                LookupFactorA(params.factor_dest_a));
+                                               
+                auto src_res = combiner_output * srcfactor;
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: Please don't use abbreviations like that. If `res` stands for "result", call it `src_result` instead.

File: src/video_core/rasterizer.cpp [hunk 8]
```diff
@@ -616,17 +616,54 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
...
                     result.r() = std::min(255, result.r());
                     result.g() = std::min(255, result.g());
                     result.b() = std::min(255, result.b());
                     combiner_output = result.Cast<u8>();
                     break;
                 }
+                
+                case params.Subtract:
+                {
+                    auto result = (src_res - dst_res) / 255;
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: Is this a Vec3<u8> or a Vec3<int>? If it's the former, the code may underflow :/

Comment (neobrain) [inline src/video_core/rasterizer.cpp]: Oh boy, this actually already affects the existing code :/

File: src/video_core/rasterizer.cpp [hunk 9]
```diff
@@ -616,17 +616,63 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
                                                LookupFactorA(params.factor_source_a));
                 auto dstfactor = Math::MakeVec(LookupFactorRGB(params.factor_dest_rgb),
                                                LookupFactorA(params.factor_dest_a));
+                                               
+                auto src_result = combiner_output * srcfactor;
+                auto dst_result = dest * dstfactor;
+
+                src_result.Cast<int>();
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: This is a nop. Sorry for not explaining this more properly; it should actually be done in the declaration itself, i.e. `auto src_result = (combiner_output * srcfactor).Cast<int>()`.

**Comment by Alegend45** [comment]
Weird, apparently, Travis can't get glfw3 anymore on OSX. Welp, that's out of the scope of my PR.

**Comment by neobrain** [comment]
Hmm interesting, thanks for the pointer.

@Alegend45 Could you double check the OpenGL spec against this? The 3DS likely does whatever that says. If not, could you make a followup PR adding a TODO to check this?

**Comment by neobrain** [comment]
(Fwiw, I guess this shows pretty well why I usually try not to have "blind" implementations lying around)

File: src/video_core/rasterizer.cpp [hunk 10]
```diff
@@ -616,17 +616,60 @@ void ProcessTriangle(const VertexShader::OutputVertex& v0,
...
+                case params.Subtract:
+                {
+                    auto result = (src_result - dst_result) / 255;
+                    result.r() = std::max(0, result.r());
+                    result.g() = std::max(0, result.g());
+                    result.b() = std::max(0, result.b());
+                    combiner_output = result.Cast<u8>();
+                    break;
+                }
+
```
Comment (neobrain) [inline src/video_core/rasterizer.cpp]: For the future, please fix your IDE not to insert trailing whitespaces all over the place ;)

Comment (Alegend45) [inline src/video_core/rasterizer.cpp]: I'm not using an IDE. Those trailing whitespaces are probably just errant
presses of the Insert key I forgot about.

**Comment by Alegend45** [comment]
#561 fixes the problems with min/max blend equations

**Pull request merged**